hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/AbstractCustomCreateOperation.java

```diff
@@ -0,0 +1,205 @@
+ *     <li>Do any Hedera-specific completion work needed when the spawned {@code CONTRACT_CREATION} succeeds.</li>
+ * </ol>
+ *
+ * <p>Note that unlike in Besu, it is possible for the second step that computes the new address to fail, since
+ * a {@code CREATE2} could reference an existing hollow account with lazy creation disabled. So this class also does
+ * a null-check after calling {@link AbstractCustomCreateOperation#setupPendingCreation(MessageFrame)}, and fails
+ * as appropriate.
+ */
+public abstract class AbstractCustomCreateOperation extends AbstractOperation {
+    private static final int MAX_STACK_DEPTH = 1024;
+    private static final Operation.OperationResult INVALID_RESPONSE =
+            new OperationResult(0L, ExceptionalHaltReason.INVALID_OPERATION);
+    private static final Operation.OperationResult UNDERFLOW_RESPONSE =
+            new Operation.OperationResult(0, ExceptionalHaltReason.INSUFFICIENT_STACK_ITEMS);
+
+    protected AbstractCustomCreateOperation(
+            final int opcode,
+            @NonNull final String name,
+            final int stackItemsConsumed,
+            final int stackItemsProduced,
+            @NonNull final GasCalculator gasCalculator) {
+        super(opcode, name, stackItemsConsumed, stackItemsProduced, gasCalculator);
+    }
+
+    /**
+     * Returns true if this operation is enabled for the given frame.
+     *
+     * @param frame the frame running the operation
+     * @return true if enabled
+     */
+    protected abstract boolean isEnabled(@NonNull MessageFrame frame);
+
+    /**
+     * Returns the gas cost of this operation for the given frame.
+     *
+     * @param frame the frame running the operation
+     * @return the gas cost
+     */
+    protected abstract long cost(@NonNull MessageFrame frame);
+
+    /**
+     * Sets up the pending creation for this operation and returns the address to be used
+     * as receiver of the child {@code CONTRACT_CREATION} message; or null if the creation
+     * should be aborted.
+     *
+     * @param frame the frame running the operation (and possibly spawning the child CONTRACT_CREATION)
+     * @return the address for the child message, or null if the creation should be aborted
+     */
+    protected abstract @Nullable Address setupPendingCreation(@NonNull final MessageFrame frame);
+
+    /**
+     * Called when the child {@code CONTRACT_CREATION} message has completed successfully,
+     * used to give specializations the chance to do any Hedera-specific logic.
+     *
+     * @param frame the frame running the successful operation
+     * @param createdAddress the address of the newly created contract
+     */
+    protected abstract void onSuccess(@NonNull MessageFrame frame, @NonNull Address createdAddress);
+
+    @Override
+    public OperationResult execute(@NonNull final MessageFrame frame, @NonNull final EVM evm) {
+        if (!isEnabled(frame)) {
```

Comment:
Did this get copied and modified from `AbstractEvmRecordingCreateOperation` or from `AbstractCreateOperation` in besu?  The implementations seems very similar but slightly different.  I'm not sure if the differences are meaningful.